sys/kern/src/arch/rv64/trap.rs

```diff
@@ -11,14 +11,149 @@ use crate::startup::with_task_table;
 use crate::task;
 use abi::{FaultInfo, FaultSource};
 
-extern crate riscv_rt;
+//use riscv::register;
+#[cfg(feature = "riscv-supervisor-mode")]
+use riscv::register::{
+    scause as xcause, scause::Exception as xcauseException,
+    scause::Interrupt as xcauseInterrupt,
+    scause::Interrupt::SupervisorTimer as xInterruptTimer,
+    scause::Trap as xcauseTrap, sepc as xepc, sscratch as xscratch,
+    stval as xtval, stvec as xtvec, stvec::TrapMode as xTrapMode,
+};
 
-use riscv::register;
-use riscv::register::mcause::{Exception, Interrupt, Trap};
+#[cfg(not(feature = "riscv-supervisor-mode"))]
+use riscv::register::{
+    mcause as xcause, mcause::Exception as xcauseException,
+    mcause::Interrupt as xcauseInterrupt,
+    mcause::Interrupt::SupervisorTimer as xInterruptTimer,
+    mcause::Trap as xcauseTrap, mepc as xepc, mscratch as xscratch,
+    mtval as xtval, mtvec as xtvec, mtvec::TrapMode as xTrapMode,
+};
 
 use crate::arch::get_current_task;
 use crate::arch::{incr_ticks, reset_timer};
 
+macro_rules! start_trap_fn_common {
+    ($scratch_reg:literal, $epc_reg:literal, $return_call:literal) => {
```

Comment:
Likewise here, I think it would be easier to follow to just using string formatting, but that might be me.  Just feels like there is 1 more layer of abstraction than is needed.  This macro could just take "s" or "m" and spit out the appropriate trap function.

Reply:
This is unfortunately complicated by the macro evaluation rules. If you attempt to use a bare macro to substitute in the prefix of an instruction mnemonic, it will not evaluate in the correct order:

`error: unrecognized instruction mnemonic
  |
note: instantiated into assembly here
 --> <inline asm>:3:13
  |
3 |             $prefixret
  |`
  
 I believe I can work around this by using paste!(...), but that's a fresh dependency.

Reply:
Fixed with concat!(...).

Reply:
Worked around using concat(...), per task.rs pattern.